File: ios/Packages/FatwaBotKit/Sources/CoreKit/OnboardingCompletionStore.swift
import Foundation

/// Tracks whether the value-first onboarding flow (docs/features/onboarding.md)
/// has been completed on this install. Local-only — onboarding runs before any
/// identity exists, so there is nothing server-side to attach this to; a
/// reinstall sees onboarding again, which is acceptable.
public struct OnboardingCompletionStore: Sendable {
    private struct Payload: Codable {
        let completed: Bool
        let completedAt: Date
    }

    private let fileURL: URL

    public init(directory: URL) {
        fileURL = directory.appendingPathComponent("onboarding-completion.json")
    }

    public func isCompleted() -> Bool {
        guard let data = try? Data(contentsOf: fileURL) else { return false }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode(Payload.self, from: data))?.completed ?? false
    }

    public func markCompleted(at date: Date = Date()) {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(Payload(completed: true, completedAt: date)) else { return }
        try? FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: android/core/common/src/main/kotlin/com/fatwabot/core/common/OnboardingCompletionStore.kt
package com.fatwabot.core.common

import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Tracks whether the value-first onboarding flow (docs/features/onboarding.md)
 * has been completed on this install. Local-only — onboarding runs before any
 * identity exists, so there is nothing server-side to attach this to; a
 * reinstall sees onboarding again, which is acceptable. Mirrors iOS
 * OnboardingCompletionStore.
 */
class OnboardingCompletionStore(private val file: File) {
    @Serializable
    private data class Payload(val completed: Boolean, val completedAtEpochSeconds: Long)

    private val json = Json { ignoreUnknownKeys = true }

    fun isCompleted(): Boolean = runCatching {
        if (!file.exists()) false else json.decodeFromString(Payload.serializer(), file.readText()).completed
    }.getOrDefault(false)

    fun markCompleted(atEpochSeconds: Long) {
        runCatching {
            file.parentFile?.mkdirs()
            val tmp = File(file.parentFile, "${file.name}.tmp")
            tmp.writeText(json.encodeToString(Payload.serializer(), Payload(true, atEpochSeconds)))
            tmp.renameTo(file)
        }
    }
}
